abdm_integrator/hiu/views/consents.py

```diff
@@ -0,0 +1,225 @@
+            consent_artefact.error = err.error
+            consent_artefact.fetch_status = ArtefactFetchStatus.ERROR
+        consent_artefact.save()
+
+    def gateway_fetch_artefact_details(self, artefact_id, payload):
+        payload['consentId'] = artefact_id
+        ABDMRequestHelper().gateway_post(HIUGatewayAPIPath.CONSENTS_FETCH, payload)
+        return payload['requestId']
+
+    def gateway_consents_on_notify(self, artefact_ids):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['acknowledgement'] = [{'status': 'OK', 'consentId': artefact_id}
+                                      for artefact_id in artefact_ids]
+        payload['resp'] = {'requestId': self.request_data['requestId']}
+        ABDMRequestHelper().gateway_post(HIUGatewayAPIPath.CONSENT_REQUEST_ON_NOTIFY, payload)
+        return payload['requestId']
+
+
+class GatewayConsentRequestOnFetch(HIUGatewayBaseView):
+
+    def post(self, request, format=None):
+        GatewayConsentRequestOnFetchSerializer(data=request.data).is_valid(raise_exception=True)
+        self.process_request(request.data)
+        return Response(status=HTTP_202_ACCEPTED)
+
+    @transaction.atomic
+    def process_request(self, request_data):
+        consent_artefact = ConsentArtefact.objects.get(gateway_request_id=request_data['resp']['requestId'])
+        if request_data.get('consent'):
+            consent_artefact.details = request_data['consent']['consentDetail']
+            consent_artefact.fetch_status = ArtefactFetchStatus.RECEIVED
+            if consent_artefact.consent_request.artefacts.filter(details__isnull=False).count() == 0:
+                self.update_consent_request_from_artefact(consent_artefact)
+        elif request_data.get('error'):
+            consent_artefact.fetch_status = ArtefactFetchStatus.ERROR
+            consent_artefact.error = request_data['error']
+        consent_artefact.save()
+
+    def update_consent_request_from_artefact(self, consent_artefact):
+        consent_request = consent_artefact.consent_request
+        consent_request.update_user_amendable_details(consent_artefact.details['permission'],
+                                                      consent_artefact.details['hiTypes'])
+
+
+class ConsentFetch(HIUBaseView, viewsets.ReadOnlyModelViewSet):
+    serializer_class = ConsentRequestSerializer
+    pagination_class = APIResultsSetPagination
+
+    def get_queryset(self):
+        queryset = ConsentRequest.objects.all().order_by('-date_created')
+        request_params = self.request.query_params
+        if request_params.get('abha_id'):
+            queryset = queryset.filter(details__patient__id=request_params['abha_id'])
+        if request_params.get('status'):
+            queryset = queryset.filter(status=request_params['status'])
+        if request_params.get('from_date'):
+            queryset = queryset.filter(health_info_to_date__date__gte=request_params['from_date'])
```

Comment:
Looks like this is filtering on to_date while using from_date value. Not sure if it is intended that way.

Reply:
Yes this is intended as a conservative filtering.
For e.g a record with date range of 1st Aug 2023 and 31st Aug 2023 will be fetched even if user provides a parameter of from date as 15th Aug 2023.